tests/assembly/s390x-vector-abi.rs

```diff
@@ -0,0 +1,255 @@
+// z10_vector-NEXT: br %r14
+// z13: vlrepg %v24, 0(%r2)
+// z13-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_ret_small(x: &i8x8) -> i8x8 {
+    *x
+}
+// CHECK-LABEL: vector_ret:
+// z10: mvc 8(8,%r2), 8(%r3)
+// z10-NEXT: mvc 0(8,%r2), 0(%r3)
+// z10-NEXT: br %r14
+// z13: vl %v24, 0(%r2), 3
+// z13-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_ret(x: &i8x16) -> i8x16 {
+    *x
+}
+// CHECK-LABEL: vector_ret_large:
+// z10: mvc 24(8,%r2), 24(%r3)
+// z10-NEXT: mvc 16(8,%r2), 16(%r3)
+// z10-NEXT: mvc 8(8,%r2), 8(%r3)
+// z10-NEXT: mvc 0(8,%r2), 0(%r3)
+// z10-NEXT: br %r14
+// z13: vl %v0, 0(%r3), 4
+// z13-NEXT: vl %v1, 16(%r3), 4
+// z13-NEXT: vst %v1, 16(%r2), 4
+// z13-NEXT: vst %v0, 0(%r2), 4
+// z13-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_ret_large(x: &i8x32) -> i8x32 {
+    *x
+}
+
+// CHECK-LABEL: vector_wrapper_ret_small:
+// CHECK: mvc 0(8,%r2), 0(%r3)
+// CHECK-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_wrapper_ret_small(x: &Wrapper<i8x8>) -> Wrapper<i8x8> {
+    *x
+}
+// CHECK-LABEL: vector_wrapper_ret:
+// CHECK: mvc 0(16,%r2), 0(%r3)
+// CHECK-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_wrapper_ret(x: &Wrapper<i8x16>) -> Wrapper<i8x16> {
+    *x
+}
+// CHECK-LABEL: vector_wrapper_ret_large:
+// z10: mvc 0(32,%r2), 0(%r3)
+// z10-NEXT: br %r14
+// z13: vl %v0, 16(%r3), 4
+// z13-NEXT: vst %v0, 16(%r2), 4
+// z13-NEXT: vl %v0, 0(%r3), 4
+// z13-NEXT: vst %v0, 0(%r2), 4
+// z13-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_wrapper_ret_large(x: &Wrapper<i8x32>) -> Wrapper<i8x32> {
+    *x
+}
+
+// CHECK-LABEL: vector_transparent_wrapper_ret_small:
+// z10: lg %r0, 0(%r3)
+// z10-NEXT: stg %r0, 0(%r2)
+// z10-NEXT: br %r14
+// z13_no_vector: lg %r0, 0(%r3)
+// z13_no_vector-NEXT: stg %r0, 0(%r2)
+// z13_no_vector-NEXT: br %r14
+// z10_vector: vlrepg %v24, 0(%r2)
+// z10_vector-NEXT: br %r14
+// z13: vlrepg %v24, 0(%r2)
+// z13-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_transparent_wrapper_ret_small(
+    x: &TransparentWrapper<i8x8>,
+) -> TransparentWrapper<i8x8> {
+    *x
+}
+// CHECK-LABEL: vector_transparent_wrapper_ret:
+// z10: mvc 8(8,%r2), 8(%r3)
+// z10-NEXT: mvc 0(8,%r2), 0(%r3)
+// z10-NEXT: br %r14
+// z13: vl %v24, 0(%r2), 3
+// z13-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_transparent_wrapper_ret(
+    x: &TransparentWrapper<i8x16>,
+) -> TransparentWrapper<i8x16> {
+    *x
+}
+// CHECK-LABEL: vector_transparent_wrapper_ret_large:
+// z10: mvc 24(8,%r2), 24(%r3)
+// z10-NEXT: mvc 16(8,%r2), 16(%r3)
+// z10-NEXT: mvc 8(8,%r2), 8(%r3)
+// z10-NEXT: mvc 0(8,%r2), 0(%r3)
+// z10-NEXT: br %r14
+// z13: vl %v0, 0(%r3), 4
+// z13-NEXT: vl %v1, 16(%r3), 4
+// z13-NEXT: vst %v1, 16(%r2), 4
+// z13-NEXT: vst %v0, 0(%r2), 4
+// z13-NEXT: br %r14
+#[no_mangle]
+extern "C" fn vector_transparent_wrapper_ret_large(
+    x: &TransparentWrapper<i8x32>,
+) -> TransparentWrapper<i8x32> {
+    *x
+}
+
+// FIXME: should check output for z10, but it is very long...
```

Comment:
It's also incorrect.  Without the vector feature, vector types of any length are supposed to be passed via implicit reference, which in the case means a pointer in %r2.  The body of the function should simply be:
```
lg %r2, 0(%r2)
```
Not sure what exactly the compiler is doing here, but it appears to have scalarized the input and pass it as 8 distinct one-byte arguments ...

Reply:
This case and soft-float case should eventually result in a compilation error once the ABI check is implemented, and I will add UI tests for these cases once the ABI check is implemented.

Reply:
Fair enough.  I just wanted to point out the incorrect code, in case we do end up going back to supporting the no-vector case.